docs/configure.rst
APRSD Configure
===============

Configure APRSD
------------------------

Once APRSD is :doc:`installed <install>` You will need to configure the config file
for running.


Generate config file
---------------------
If you have never run the server, running it the first time will generate
a sample config file in the default location of ~/.config/aprsd/aprsd.yml

.. code-block:: shell

    └─> aprsd server
    12/28/2022 04:26:31 PM MainThread           ERROR    No config file found!! run 'aprsd sample-config'                             cli_helper.py:90
    12/28/2022 04:26:31 PM MainThread           ERROR    Config aprs_network.password not set.                                           client.py:105
    12/28/2022 04:26:31 PM MainThread           ERROR    Option 'aprs_network.password is not set.' was not in config file               client.py:268
    12/28/2022 04:26:31 PM MainThread           ERROR    APRS client is not properly configured in config file.                           server.py:58

You can see the sample config file output

Sample config file
------------------

.. code-block:: shell

    └─> aprsd sample-config
    [DEFAULT]

    #
    # From aprsd.conf
    #

    # Callsign to use for messages sent by APRSD (string value)
    #callsign = <None>

    # Enable saving of watch list, packet tracker between restarts.
    # (boolean value)
    #enable_save = true

    # Save location for packet tracking files. (string value)
    #save_location = ~/.config/aprsd

    # Enable code tracing (boolean value)
    #trace_enabled = false

    # Units for display, imperial or metric (string value)
    #units = imperial

    # Comma separated list of enabled plugins for APRSD.To enable
    # installed external plugins add them here.The full python path to the
    # class name must be used (list value)
    #enabled_plugins = aprsd.plugins.email.EmailPlugin,aprsd.plugins.fortune.FortunePlugin,aprsd.plugins.location.LocationPlugin,aprsd.plugins.ping.PingPlugin,aprsd.plugins.query.QueryPlugin,aprsd.plugins.time.TimePlugin,aprsd.plugins.weather.OWMWeatherPlugin,aprsd.plugins.version.VersionPlugin


    [admin]

    #
    # From aprsd.conf
    #

    # Enable the Admin Web Interface (boolean value)
    #web_enabled = false

    # The ip address to listen on (IP address value)
    #web_ip = 0.0.0.0

    # The port to listen on (port value)
    # Minimum value: 0
    # Maximum value: 65535
    #web_port = 8001

    # The admin user for the admin web interface (string value)
    #user = admin

    # Admin interface password (string value)
    #password = <None>


    [aprs_fi]

    #
    # From aprsd.conf
    #

    # Get the apiKey from your aprs.fi account here:http://aprs.fi/account
    # (string value)
    #apiKey = <None>


    [aprs_network]

    #
    # From aprsd.conf
    #

    # Set enabled to False if there is no internet connectivity.This is
    # useful for a direwolf KISS aprs connection only. (boolean value)
    #enabled = true

    # APRS Username (string value)
    #login = NOCALL

    # APRS Password Get the passcode for your callsign here:
    # https://apps.magicbug.co.uk/passcode (string value)
    #password = <None>

    # The APRS-IS hostname (hostname value)
    #host = noam.aprs2.net

    # APRS-IS port (port value)
    # Minimum value: 0
    # Maximum value: 65535
    #port = 14580


    [aprsd_weewx_plugin]

    #
    # From aprsd_weewx_plugin.conf
    #

    # Latitude of the station you want to report as (floating point value)
    #latitude = <None>

    # Longitude of the station you want to report as (floating point
    # value)
    #longitude = <None>

    # How long (in seconds) in between weather reports (integer value)
    #report_interval = 60


    [avwx_plugin]

    #
    # From aprsd.conf
    #

    # avwx-api is an opensource project that hasa hosted service here:
    # https://avwx.rest/You can launch your own avwx-api in a containerby
    # cloning the githug repo here:https://github.com/avwx-rest/AVWX-API
    # (string value)
    #apiKey = <None>

    # The base url for the avwx API.  If you are hosting your ownHere is
    # where you change the url to point to yours. (string value)
    #base_url = https://avwx.rest


    [email_plugin]

    #
    # From aprsd.conf
    #

    # (Required) Callsign to validate for doing email commands.Only this
    # callsign can check email. This is also where the email notifications
    # for new emails will be sent. (string value)
    #callsign = <None>

    # Enable the Email plugin? (boolean value)
    #enabled = false

    # Enable the Email plugin Debugging? (boolean value)
    #debug = false

    # Login username/email for IMAP server (string value)
    #imap_login = <None>

    # Login password for IMAP server (string value)
    #imap_password = <None>

    # Hostname/IP of the IMAP server (hostname value)
    #imap_host = <None>

    # Port to use for IMAP server (port value)
    # Minimum value: 0
    # Maximum value: 65535
    #imap_port = 993

    # Use SSL for connection to IMAP Server (boolean value)
    #imap_use_ssl = true

    # Login username/email for SMTP server (string value)
    #smtp_login = <None>

    # Login password for SMTP server (string value)
    #smtp_password = <None>

    # Hostname/IP of the SMTP server (hostname value)
    #smtp_host = <None>

    # Port to use for SMTP server (port value)
    # Minimum value: 0
    # Maximum value: 65535
    #smtp_port = 465

    # Use SSL for connection to SMTP Server (boolean value)
    #smtp_use_ssl = true

    # List of email shortcuts for checking/sending email For Exmaple:
    # wb=walt@walt.com,cl=cl@cl.com
    # Means use 'wb' to send an email to walt@walt.com (list value)
    #email_shortcuts = <None>


    [kiss_serial]

    #
    # From aprsd.conf
    #

    # Enable Serial KISS interface connection. (boolean value)
    #enabled = false

    # Serial Device file to use.  /dev/ttyS0 (string value)
    #device = <None>

    # The Serial device baud rate for communication (integer value)
    #baudrate = 9600


    [kiss_tcp]

    #
    # From aprsd.conf
    #

    # Enable Serial KISS interface connection. (boolean value)
    #enabled = false

    # The KISS TCP Host to connect to. (hostname value)
    #host = <None>

    # The KISS TCP/IP network port (port value)
    # Minimum value: 0
    # Maximum value: 65535
    #port = 8001


    [logging]

    #
    # From aprsd.conf
    #

    # Date format for log entries (string value)
    #date_format = %m/%d/%Y %I:%M:%S %p

    # Enable Rich logging (boolean value)
    #rich_logging = true

    # File to log to (string value)
    #logfile = <None>

    # Log file format, unless rich_logging enabled. (string value)
    #logformat = [%(asctime)s] [%(threadName)-20.20s] [%(levelname)-5.5s] %(message)s - [%(pathname)s:%(lineno)d]


    [owm_weather_plugin]

    #
    # From aprsd.conf
    #

    # OWMWeatherPlugin api key to OpenWeatherMap's API.This plugin uses
    # the openweathermap API to fetchlocation and weather information.To
    # use this plugin you need to get an openweathermapaccount and
    # apikey.https://home.openweathermap.org/api_keys (string value)
    #apiKey = <None>


    [query_plugin]

    #
    # From aprsd.conf
    #

    # The Ham callsign to allow access to the query plugin from RF.
    # (string value)
    #callsign = <None>


    [rpc_settings]

    #
    # From aprsd.conf
    #

    # Enable RPC calls (boolean value)
    #enabled = true

    # The ip address to listen on (string value)
    #ip = localhost

    # The port to listen on (port value)
    # Minimum value: 0
    # Maximum value: 65535
    #port = 18861

    # Magic word to authenticate requests between client/server (string
    # value)
    #magic_word = CHANGEME!!!


    [watch_list]

    #
    # From aprsd.conf
    #

    # Enable the watch list feature.  Still have to enable the correct
    # plugin.  Built-in plugin to use is aprsd.plugins.notify.NotifyPlugin
    # (boolean value)
    #enabled = false

    # Callsigns to watch for messsages (list value)
    #callsigns = <None>

    # The Ham Callsign to send messages to for watch list alerts. (string
    # value)
    #alert_callsign = <None>

    # The number of packets to store. (integer value)
    #packet_keep_count = 10

    # Time to wait before alert is sent on new message for users in
    # callsigns. (integer value)
    #alert_time_seconds = 3600

Note, You must edit the config file and change the ham callsign to your
legal FCC HAM callsign, or aprsd server will not start.

.. include:: links.rst


ChangeLog
CHANGES
=======

* Decouple admin web interface from server command
* Dockerfile now produces aprsd.conf
* Fix some unit tests and loading of CONF w/o file
* Added missing conf
* Removed references to old custom config
* Convert config to oslo\_config
* Added rain formatting unit tests to WeatherPacket
* Fix Rain reporting in WeatherPacket send
* Removed Packet.send()
* Removed watchlist plugins
* Fix PluginManager.get\_plugins
* Cleaned up PluginManager
* Cleaned up PluginManager
* Update routing for weatherpacket
* Fix some WeatherPacket formatting
* Fix pep8 violation
* Add packet filtering for aprsd listen
* Added WeatherPacket encoding
* Updated webchat and listen for queue based RX
* reworked collecting and reporting stats
* Removed unused threading code
* Change RX packet processing to enqueu
* Make tracking objectstores work w/o initializing
* Cleaned up packet transmit class attributes
* Fix packets timestamp to int
* More messaging -> packets cleanup
* Cleaned out all references to messaging
* Added contructing a GPSPacket for sending
* cleanup webchat
* Reworked all packet processing
* Updated plugins and plugin interfaces for Packet
* Started using dataclasses to describe packets

v2.6.1
------

* v2.6.1
* Fixed position report for webchat beacon
* Try and fix broken 32bit qemu builds on 64bit system
* Add unit tests for webchat
* remove armv7 build RUST sucks
* Fix for Collections change in 3.10

v2.6.0
------

* Update workflow again
* Update Dockerfile to 22.04
* Update Dockerfile and build.sh
* Update workflow
* Prep for 2.6.0 release
* Update requirements
* Removed Makefile comment
* Update Makefile for dev vs. run environments
* Added pyopenssl for https for webchat
* change from device-detector to user-agents
* Remove twine from dev-requirements
* Update to latest Makefile.venv
* Refactored threads a bit
* Mark packets as acked in MsgTracker
* remove dev setting for template
* Add GPS beacon to mobile page
* Allow werkzeug for admin interface
* Allow werkzeug for admin interface
* Add support for mobile browsers for webchat
* Ignore callsign case while processing packets
* remove linux/arm/v7 for official builds for now
* added workflow for building specific version
* Allow passing in version to the Dockerfile
* Send GPS Beacon from webchat interface
* specify Dockerfile-dev
* Fixed build.sh
* Build on the source not released aprsd
* Remove email validation
* Add support for building linux/arm/v7
* Remove python 3.7 from docker build github
* Fixed failing unit tests
* change github workflow
* Removed TimeOpenCageDataPlugin
* Dump config with aprsd dev test-plugin
* Updated requirements
* Got webchat working with KISS tcp
* Added click auto\_envvar\_prefix
* Update aprsd thread base class to use queue
* Update packets to use wrapt
* Add remving existing requirements
* Try sending raw APRSFrames to aioax25
* Use new aprsd.callsign as the main callsign
* Fixed access to threads refactor
* Added webchat command
* Moved log.py to logging
* Moved trace.py to utils
* Fixed pep8 errors
* Refactored threads.py
* Refactor utils to directory
* remove arm build for now
* Added rustc and cargo to Dockerfile
* remove linux/arm/v6 from docker platform build
* Only tag master build as master
* Remove docker build from test
* create master-build.yml
* Added container build action
* Update docs on using Docker
* Update dev-requirements pip-tools
* Fix typo in docker-compose.yml
* Fix PyPI scraping
* Allow web interface when running in Docker
* Fix typo on exception
* README formatting fixes
* Bump dependencies to fix python 3.10
* Fixed up config option checking for KISS
* Fix logging issue with log messages
* for 2.5.9

v2.5.9
------

* FIX: logging exceptions
* Updated build and run for rich lib
* update build for 2.5.8

v2.5.8
------

* For 2.5.8
* Removed debug code
* Updated list-plugins
* Renamed virtualenv dir to .aprsd-venv
* Added unit tests for dev test-plugin
* Send Message command defaults to config

v2.5.7
------

* Updated Changelog
* Fixed an KISS config disabled issue
* Fixed a bug with multiple notify plugins enabled
* Unify the logging to file and stdout
* Added new feature to list-plugins command
* more README.rst cleanup
* Updated README examples

v2.5.6
------

* Changelog
* Tightened up the packet logging
* Added unit tests for USWeatherPlugin, USMetarPlugin
* Added test\_location to test LocationPlugin
* Updated pytest output
* Added py39 to tox for tests
* Added NotifyPlugin unit tests and more
* Small cleanup on packet logging
* Reduced the APRSIS connection reset to 2 minutes
* Fixed the NotifyPlugin
* Fixed some pep8 errors
* Add tracing for dev command
* Added python rich library based logging
* Added LOG\_LEVEL env variable for the docker

v2.5.5
------

* Update requirements to use aprslib 0.7.0
* fixed the failure during loading for objectstore
* updated docker build

v2.5.4
------

* Updated Changelog
* Fixed dev command missing initialization

v2.5.3
------

* Fix admin logging tab

v2.5.2
------

* Added new list-plugins command
* Don't require check-version command to have a config
* Healthcheck command doesn't need the aprsd.yml config
* Fix test failures
* Removed requirement for aprs.fi key
* Updated Changelog

v2.5.1
------

* Removed stock plugin
* Removed the stock plugin

v2.5.0
------

* Updated for v2.5.0
* Updated Dockerfile's and build script for docker
* Cleaned up some verbose output & colorized output
* Reworked all the common arguments
* Fixed test-plugin
* Ensure common params are honored
* pep8
* Added healthcheck to the cmds
* Removed the need for FROMCALL in dev test-plugin
* Pep8 failures
* Refactor the cli
* Updated Changelog for 4.2.3
* Fixed a problem with send-message command

v2.4.2
------

* Updated Changelog
* Be more careful picking data to/from disk
* Updated Changelog

v2.4.1
------

* Ensure plugins are last to be loaded
* Fixed email connecting to smtp server

v2.4.0
------

* Updated Changelog for 2.4.0 release

aprsd/config.py
import collections
import logging
import os
from pathlib import Path
import sys

import click
import yaml

from aprsd import exception, utils


home = str(Path.home())
DEFAULT_CONFIG_DIR = f"{home}/.config/aprsd/"
DEFAULT_SAVE_FILE = f"{home}/.config/aprsd/aprsd.p"
DEFAULT_CONFIG_FILE = f"{home}/.config/aprsd/aprsd.yml"

LOG_LEVELS = {
    "CRITICAL": logging.CRITICAL,
    "ERROR": logging.ERROR,
    "WARNING": logging.WARNING,
    "INFO": logging.INFO,
    "DEBUG": logging.DEBUG,
}

DEFAULT_DATE_FORMAT = "%m/%d/%Y %I:%M:%S %p"
DEFAULT_LOG_FORMAT = (
    "[%(asctime)s] [%(threadName)-20.20s] [%(levelname)-5.5s]"
    " %(message)s - [%(pathname)s:%(lineno)d]"
)

QUEUE_DATE_FORMAT = "[%m/%d/%Y] [%I:%M:%S %p]"
QUEUE_LOG_FORMAT = (
    "%(asctime)s [%(threadName)-20.20s] [%(levelname)-5.5s]"
    " %(message)s - [%(pathname)s:%(lineno)d]"
)

CORE_MESSAGE_PLUGINS = [
    "aprsd.plugins.email.EmailPlugin",
    "aprsd.plugins.fortune.FortunePlugin",
    "aprsd.plugins.location.LocationPlugin",
    "aprsd.plugins.ping.PingPlugin",
    "aprsd.plugins.query.QueryPlugin",
    "aprsd.plugins.time.TimePlugin",
    "aprsd.plugins.weather.USWeatherPlugin",
    "aprsd.plugins.version.VersionPlugin",
]

CORE_NOTIFY_PLUGINS = [
    "aprsd.plugins.notify.NotifySeenPlugin",
]

ALL_PLUGINS = []
for i in CORE_MESSAGE_PLUGINS:
    ALL_PLUGINS.append(i)
for i in CORE_NOTIFY_PLUGINS:
    ALL_PLUGINS.append(i)

# an example of what should be in the ~/.aprsd/config.yml
DEFAULT_CONFIG_DICT = {
    "ham": {"callsign": "NOCALL"},
    "aprs": {
        "enabled": True,
        # Only used as the login for aprsis.
        "login": "CALLSIGN",
        "password": "00000",
        "host": "rotate.aprs2.net",
        "port": 14580,
    },
    "kiss": {
        "tcp": {
            "enabled": False,
            "host": "direwolf.ip.address",
            "port": "8001",
        },
        "serial": {
            "enabled": False,
            "device": "/dev/ttyS0",
            "baudrate": 9600,
        },
    },
    "aprsd": {
        # Callsign to use for all packets to/from aprsd instance
        # regardless of the client (aprsis vs kiss)
        "callsign": "NOCALL",
        "logfile": "/tmp/aprsd.log",
        "logformat": DEFAULT_LOG_FORMAT,
        "dateformat": DEFAULT_DATE_FORMAT,
        "save_location": DEFAULT_CONFIG_DIR,
        "rich_logging": True,
        "trace": False,
        "enabled_plugins": ALL_PLUGINS,
        "units": "imperial",
        "watch_list": {
            "enabled": False,
            # Who gets the alert?
            "alert_callsign": "NOCALL",
            # 43200 is 12 hours
            "alert_time_seconds": 43200,
            # How many packets to save in a ring Buffer
            # for a particular callsign
            "packet_keep_count": 10,
            "callsigns": [],
        },
        "web": {
            "enabled": True,
            "logging_enabled": True,
            "host": "0.0.0.0",
            "port": 8001,
            "users": {
                "admin": "password-here",
            },
        },
        "email": {
            "enabled": True,
            "shortcuts": {
                "aa": "5551239999@vtext.com",
                "cl": "craiglamparter@somedomain.org",
                "wb": "555309@vtext.com",
            },
            "smtp": {
                "login": "SMTP_USERNAME",
                "password": "SMTP_PASSWORD",
                "host": "smtp.gmail.com",
                "port": 465,
                "use_ssl": False,
                "debug": False,
            },
            "imap": {
                "login": "IMAP_USERNAME",
                "password": "IMAP_PASSWORD",
                "host": "imap.gmail.com",
                "port": 993,
                "use_ssl": True,
                "debug": False,
            },
        },
    },
    "services": {
        "aprs.fi": {"apiKey": "APIKEYVALUE"},
        "openweathermap": {"apiKey": "APIKEYVALUE"},
        "opencagedata": {"apiKey": "APIKEYVALUE"},
        "avwx": {"base_url": "http://host:port", "apiKey": "APIKEYVALUE"},
    },
}


class Config(collections.UserDict):
    def _get(self, d, keys, default=None):
        """
        Example:
            d = {'meta': {'status': 'OK', 'status_code': 200}}
            _get(d, ['meta', 'status_code'])          # => 200
            _get(d, ['garbage', 'status_code'])       # => None
            _get(d, ['meta', 'garbage'], default='-') # => '-'

        """
        if type(keys) is str and "." in keys:
            keys = keys.split(".")

        assert type(keys) is list
        if d is None:
            return default

        if not keys:
            return d

        if type(d) is str:
            return default

        return self._get(d.get(keys[0]), keys[1:], default)

    def get(self, path, default=None):
        return self._get(self.data, path, default=default)

    def exists(self, path):
        """See if a conf value exists."""
        test = "-3.14TEST41.3-"
        return self.get(path, default=test) != test

    def check_option(self, path, default_fail=None):
        """Make sure the config option doesn't have default value."""
        if not self.exists(path):
            if type(path) is list:
                path = ".".join(path)
            raise exception.MissingConfigOptionException(path)

        val = self.get(path)
        if val == default_fail:
            # We have to fail and bail if the user hasn't edited
            # this config option.
            raise exception.ConfigOptionBogusDefaultException(
                path, default_fail,
            )


def add_config_comments(raw_yaml):
    end_idx = utils.end_substr(raw_yaml, "ham:")
    if end_idx != -1:
        # lets insert a comment
        raw_yaml = utils.insert_str(
            raw_yaml,
            "\n    # Callsign that owns this instance of APRSD.",
            end_idx,
        )
    end_idx = utils.end_substr(raw_yaml, "aprsd:")
    if end_idx != -1:
        # lets insert a comment
        raw_yaml = utils.insert_str(
            raw_yaml,
            "\n    # Callsign to use for all APRSD Packets as the to/from."
            "\n    # regardless of client type (aprsis vs tcpkiss vs serial)",
            end_idx,
        )
    end_idx = utils.end_substr(raw_yaml, "aprs:")
    if end_idx != -1:
        # lets insert a comment
        raw_yaml = utils.insert_str(
            raw_yaml,
            "\n    # Set enabled to False if there is no internet connectivity."
            "\n    # This is useful for a direwolf KISS aprs connection only. "
            "\n"
            "\n    # Get the passcode for your callsign here: "
            "\n    # https://apps.magicbug.co.uk/passcode",
            end_idx,
        )

    end_idx = utils.end_substr(raw_yaml, "aprs.fi:")
    if end_idx != -1:
        # lets insert a comment
        raw_yaml = utils.insert_str(
            raw_yaml,
            "\n        # Get the apiKey from your aprs.fi account here:  "
            "\n        # http://aprs.fi/account",
            end_idx,
        )

    end_idx = utils.end_substr(raw_yaml, "opencagedata:")
    if end_idx != -1:
        # lets insert a comment
        raw_yaml = utils.insert_str(
            raw_yaml,
            "\n        # (Optional for TimeOpenCageDataPlugin) "
            "\n        # Get the apiKey from your opencagedata account here:  "
            "\n        # https://opencagedata.com/dashboard#api-keys",
            end_idx,
        )

    end_idx = utils.end_substr(raw_yaml, "openweathermap:")
    if end_idx != -1:
        # lets insert a comment
        raw_yaml = utils.insert_str(
            raw_yaml,
            "\n        # (Optional for OWMWeatherPlugin) "
            "\n        # Get the apiKey from your "
            "\n        # openweathermap account here: "
            "\n        # https://home.openweathermap.org/api_keys",
            end_idx,
        )

    end_idx = utils.end_substr(raw_yaml, "avwx:")
    if end_idx != -1:
        # lets insert a comment
        raw_yaml = utils.insert_str(
            raw_yaml,
            "\n        # (Optional for AVWXWeatherPlugin) "
            "\n        # Use hosted avwx-api here: https://avwx.rest "
            "\n        # or deploy your own from here: "
            "\n        # https://github.com/avwx-rest/avwx-api",
            end_idx,
        )

    return raw_yaml


def dump_default_cfg():
    return add_config_comments(
        yaml.dump(
            DEFAULT_CONFIG_DICT,
            indent=4,
        ),
    )


def create_default_config():
    """Create a default config file."""
    # make sure the directory location exists
    config_file_expanded = os.path.expanduser(DEFAULT_CONFIG_FILE)
    config_dir = os.path.dirname(config_file_expanded)
    if not os.path.exists(config_dir):
        click.echo(f"Config dir '{config_dir}' doesn't exist, creating.")
        utils.mkdir_p(config_dir)
    with open(config_file_expanded, "w+") as cf:
        cf.write(dump_default_cfg())


def get_config(config_file):
    """This tries to read the yaml config from <config_file>."""
    config_file_expanded = os.path.expanduser(config_file)
    if os.path.exists(config_file_expanded):
        with open(config_file_expanded) as stream:
            config = yaml.load(stream, Loader=yaml.FullLoader)
            return Config(config)
    else:
        if config_file == DEFAULT_CONFIG_FILE:
            click.echo(
                f"{config_file_expanded} is missing, creating config file",
            )
            create_default_config()
            msg = (
                "Default config file created at {}.  Please edit with your "
                "settings.".format(config_file)
            )
            click.echo(msg)
        else:
            # The user provided a config file path different from the
            # Default, so we won't try and create it, just bitch and bail.
            msg = f"Custom config file '{config_file}' is missing."
            click.echo(msg)

        sys.exit(-1)


# This method tries to parse the config yaml file
# and consume the settings.
# If the required params don't exist,
# it will look in the environment
def parse_config(config_file):
    config = get_config(config_file)

    def fail(msg):
        click.echo(msg)
        sys.exit(-1)

    def check_option(config, path, default_fail=None):
        try:
            config.check_option(path, default_fail=default_fail)
        except Exception as ex:
            fail(repr(ex))
        else:
            return config

    # special check here to make sure user has edited the config file
    # and changed the ham callsign
    check_option(
        config,
        "ham.callsign",
        default_fail=DEFAULT_CONFIG_DICT["ham"]["callsign"],
    )
    check_option(
        config,
        ["aprsd"],
    )
    check_option(
        config,
        "aprsd.callsign",
        default_fail=DEFAULT_CONFIG_DICT["aprsd"]["callsign"],
    )

    # Ensure they change the admin password
    if config.get("aprsd.web.enabled") is True:
        check_option(
            config,
            ["aprsd", "web", "users", "admin"],
            default_fail=DEFAULT_CONFIG_DICT["aprsd"]["web"]["users"]["admin"],
        )

    if config.get("aprsd.watch_list.enabled") is True:
        check_option(
            config,
            ["aprsd", "watch_list", "alert_callsign"],
            default_fail=DEFAULT_CONFIG_DICT["aprsd"]["watch_list"]["alert_callsign"],
        )

    if config.get("aprsd.email.enabled") is True:
        # Check IMAP server settings
        check_option(config, ["aprsd", "email", "imap", "host"])
        check_option(config, ["aprsd", "email", "imap", "port"])
        check_option(
            config,
            ["aprsd", "email", "imap", "login"],
            default_fail=DEFAULT_CONFIG_DICT["aprsd"]["email"]["imap"]["login"],
        )
        check_option(
            config,
            ["aprsd", "email", "imap", "password"],
            default_fail=DEFAULT_CONFIG_DICT["aprsd"]["email"]["imap"]["password"],
        )

        # Check SMTP server settings
        check_option(config, ["aprsd", "email", "smtp", "host"])
        check_option(config, ["aprsd", "email", "smtp", "port"])
        check_option(
            config,
            ["aprsd", "email", "smtp", "login"],
            default_fail=DEFAULT_CONFIG_DICT["aprsd"]["email"]["smtp"]["login"],
        )
        check_option(
            config,
            ["aprsd", "email", "smtp", "password"],
            default_fail=DEFAULT_CONFIG_DICT["aprsd"]["email"]["smtp"]["password"],
        )

    return config
